From 5b03ae6305ab919153b9102ab3a3fe16b10be4ec Mon Sep 17 00:00:00 2001
From: Christine Gharzuzi <cgharzuzi@habana.ai>
Date: Tue, 08 Feb 2022 19:52:55 +0200
Subject: [PATCH] [SW-63244] gaudi2: jump to __reset address of zephyr preboot

Similar to running PID/MGMT on ARCs, the __reset address
of the binary is availablein the first entry of the
preboot.bin, therefore dereference is needed.

Change-Id: I5d42e657b75f406bbbeb1d6b17a8149b07d807f2
Signed-off-by: Christine Gharzuzi <cgharzuzi@habana.ai>
---

diff --git a/arch/gaudi2/arc/common.c b/arch/gaudi2/arc/common.c
index 8dac81d..c4c7997 100644
--- a/arch/gaudi2/arc/common.c
+++ b/arch/gaudi2/arc/common.c
@@ -162,7 +162,13 @@
 void image_start(u64 loadaddr)
 {
 	u32 lsb = (u32) loadaddr;
-	void (*image_entry)(void) = (void *) lsb;
+	void (*image_entry)(void);
+
+	/*
+	 * Jump to __reset of zephyr. After image_entry, control is not
+	 * expected to return here and continue.
+	 */
+	image_entry = (void *) (*(u32 *) lsb);
 
 	image_entry();
 }
